### DISCUSSION TITLE
C# Winform: Slide show is not working when i use Quartz.net to fire my routine at specific time

### DISCUSSION THREAD

**Discussion by karlosRivera** [quartznet/quartznet#451] [General] [answered]
here i am giving a small my code snippet just to show what i am trying to achieve.
i guess quartz.net creating issue for me.

my objective is to call my routine every day at specific time and which will play slide show. the problem is slide show image is not coming when i invoke my routine with quartz.net and when i am not using quartz.net then everything works fine but i have to use quartz.net to schedule my routine.

private void frmMain_Load(object sender, EventArgs e)
{
    // construct a scheduler factory
    ISchedulerFactory schedFact = new StdSchedulerFactory();

    // get a scheduler
    sched = schedFact.GetScheduler();
    sched.Start();

    IJobDetail job = JobBuilder.Create<frmMain>()
        .WithIdentity("Job", "group")
        .Build();

    ITrigger trigger = TriggerBuilder.Create()
       .WithDailyTimeIntervalSchedule
         (s =>
            s.WithIntervalInHours(24)
           .OnEveryDay()
           .StartingDailyAt(TimeOfDay.HourAndMinuteOfDay(01, 55))
         )
       .Build();

    sched.ScheduleJob(job, trigger);
}

public void Execute(IJobExecutionContext context)
{
    generate();
}

public void generate()
{
    if (this.FetchStart != null)
        this.FetchStart(this, new EventArgs());

    System.Threading.Thread.Sleep(5000);

    if (this.FetchDone != null)
        this.FetchDone(this, new EventArgs());
}

i have done my project with VS2013 community edition. my objective is to call slide show routine every day at specific time. when i am calling my slide show routine without quartz.net scheduler then it is working fine but when i invoke my routine by quartz.net scheduler then routine is getting called but no slide show image is showing.

what [problem occur is not clear to me. as per my objective i have to use quartz.net scheduler because i need to invoke my routine at a specific time of day every day.

here i am sharing my project code because it is in onedrive. so my request please some one download my project and run at your end to see the problem and tell me the reason which causes not to show images on picture box.

if possible please rectify my code with quartz.net scheduler code. one drive project link is https://1drv.ms/f/s!AmIfMNV-CodPa81zFiNH6Ur7qro

i upload my project folder.

thanks
UPDATE

when i use background worker along with quartz.net to call my generate routine then also no improvement i found. same problem that slide show image is not appearing on picture box.
here is the code for quartz.net with background worker

public partial class frmMain : Form, IJob
    {
        ucSlide oSlide = new ucSlide();
        IScheduler sched = null;
        BackgroundWorker m_oWorker = null;

        public event EventHandler FetchStart;
        public event EventHandler FetchDone;
        public event EventHandler NoDataFound;

        public frmMain()
        {
            InitializeComponent();

            m_oWorker = new BackgroundWorker();
            m_oWorker.DoWork += new DoWorkEventHandler(m_oWorker_DoWork);
            m_oWorker.RunWorkerCompleted += new RunWorkerCompletedEventHandler(m_oWorker_RunWorkerCompleted);

            oSlide.MainForm = this;
            oSlide.SlideSource = Utility.SlidePath;

            this.Controls.Add(oSlide);
            oSlide.Dock = DockStyle.Fill;
        }

        private void frmMain_Load(object sender, EventArgs e)
        {
            // construct a scheduler factory
            ISchedulerFactory schedFact = new StdSchedulerFactory();

            // get a scheduler
            sched = schedFact.GetScheduler();
            sched.Start();

            IJobDetail job = JobBuilder.Create<frmMain>()
                .WithIdentity("Job", "group")
                .Build();

            ITrigger trigger = TriggerBuilder.Create()
               .WithDailyTimeIntervalSchedule
                 (s =>
                    s.WithIntervalInHours(24)
                   .OnEveryDay()
                   .StartingDailyAt(TimeOfDay.HourAndMinuteOfDay(02,23))
                 )
               .Build();

            sched.ScheduleJob(job, trigger);
        }

        public void Execute(IJobExecutionContext context)
        {
            m_oWorker.RunWorkerAsync();

        }

        public void generate()
        {
            if (this.FetchStart != null)
                this.FetchStart(this, new EventArgs());

            System.Threading.Thread.Sleep(5000);

            if (this.FetchDone != null)
                this.FetchDone(this, new EventArgs());
        }

        void m_oWorker_DoWork(object sender, DoWorkEventArgs e)
        {
            generate();
        }

        void m_oWorker_RunWorkerCompleted(object sender, RunWorkerCompletedEventArgs e)
        {
            // The background process is complete. We need to inspect
            // our response to see if an error occurred, a cancel was
            // requested or if we completed successfully.  
            //if (e.Cancelled)
            //{
            //    //lblStatus.Text = "Task Cancelled.";
            //    isBusy = false;
            //}

            //// Check to see if an error occurred in the background process.

            //else if (e.Error != null)
            //{
            //    //lblStatus.Text = "Error while performing background operation.";
            //    MessageBox.Show(e.Error.InnerException.Message.ToString());
            //    isBusy = false;
            //}
            //else
            //{
            //    // Everything completed normally.
            //    //lblStatus.Text = "Task Completed...";
            //    isBusy = false;
            //}
        }
    }

thanks

**Comment by Liwoj** [comment]
Your example doesn't contain code used to update UI so its a guess but from the description i think problem could be Windows Forms thread affinity (BackgroundWorker uses thread pool and Quartz.NET is using its own thread pool for job execution).
Take a look and [this arcicle](https://msdn.microsoft.com/en-us/library/ms171728(v=vs.110).aspx) on MSDN on how to update UI from other thread then your app main thread...

**Reply by mcarbonneaux** [reply]
effectively when i've call my method from quartz jobs through delegate it's worked fine.

i've made static the instantiation of my form in main.
```
static class Program
  {
    public static My_form oMy_form= null;

    /// <summary>
    /// Point d'entrée principal de l'application.
    /// </summary>
    [STAThread]
    static void Main()
    {
      
      Application.EnableVisualStyles();
      Application.SetCompatibleTextRenderingDefault(false);
      
      oMy_form = new My_form();
      
      Application.Run(oMy_form);
    }
  }
```
i've added method on my form to be called from my quartz job.
in my quartz jobs i call this method from static varriable.
```
  class myJobs : IJob
  {
    public Task Execute(IJobExecutionContext context)
    {
      Program.oMy_form.mymethod_in_my_form();
      return Task.CompletedTask;
    }
  }
```
in this method i check invokerequired, them call my with delagated.

```
    private delegate void mymethod_in_my_form_Delegate();
    public void mymethod_in_my_form()
    {
      
      if (this.InvokeRequired)
      {
        var d = new mymethod_in_my_form_Delegate(mymethod_in_my_form);
        this.Invoke(d);
      }
      // do all your call to control of the form your in the thread of the form
   }
```

**Comment by karlosRivera** [comment, marked as answer]
after spending lots of hour i found out the way to get my code work.

i have to pass my form reference with scheduled this way.

    ISchedulerFactory schedFact = new StdSchedulerFactory();

    // get a scheduler
    sched = schedFact.GetScheduler();
    sched.Start();

    IJobDetail job = JobBuilder.Create<frmMain>()
        .WithIdentity("Job", "group")
        .Build();

    job.JobDataMap.Put("Form", this);

    ITrigger trigger = TriggerBuilder.Create()
       .WithDailyTimeIntervalSchedule
         (s =>
            s.WithIntervalInHours(24)
           .OnEveryDay()
           .StartingDailyAt(TimeOfDay.HourAndMinuteOfDay(20, 27))
         )
       .Build();

    sched.ScheduleJob(job, trigger);

the main trick lies here

job.JobDataMap.Put("Form", this);

AND

public void Execute(IJobExecutionContext context)
{
    var dataMap = context.MergedJobDataMap;
    var frmInstance = (frmMain) dataMap["Form"];
    frmInstance.generate();
}

all thanks goes to @hillstuk and his link is http://stackoverflow.com/a/25660846/728750

my full code slide show with quartz.net

using System;
using System.IO;
using System.Windows.Forms;
using Quartz;
using Quartz.Impl;
using System.Timers;

namespace SchedulerWithTimer
{
    public partial class frmMain : Form, IJob
    {
        private int imageNumber = 1, SlideCount = 0;
        string _SlideSource = "";
        System.Timers.Timer aTimer = new System.Timers.Timer();

        IScheduler sched = null;

        public frmMain()
        {
            InitializeComponent();
            _SlideSource = Utility.SlidePath;
            SlideCount = Directory.GetFiles(_SlideSource, "*", SearchOption.AllDirectories).Length;

            aTimer.Elapsed += new ElapsedEventHandler(OnTimedEvent);
            aTimer.Interval = 5000;

            aTimer.Enabled = false;
        }

        private void frmMain_Load(object sender, EventArgs e)
        {
            // construct a scheduler factory
            ISchedulerFactory schedFact = new StdSchedulerFactory();

            // get a scheduler
            sched = schedFact.GetScheduler();
            sched.Start();

            IJobDetail job = JobBuilder.Create<frmMain>()
                .WithIdentity("Job", "group")
                .Build();

            job.JobDataMap.Put("Form", this);

            ITrigger trigger = TriggerBuilder.Create()
               .WithDailyTimeIntervalSchedule
                 (s =>
                    s.WithIntervalInHours(24)
                   .OnEveryDay()
                   .StartingDailyAt(TimeOfDay.HourAndMinuteOfDay(20, 27))
                 )
               .Build();

            sched.ScheduleJob(job, trigger);
        }

        public void Execute(IJobExecutionContext context)
        {
            var dataMap = context.MergedJobDataMap;
            var frmInstance = (frmMain) dataMap["Form"];
            frmInstance.generate();
        }


        public void generate()
        {
            FetchStart();
            System.Threading.Thread.Sleep(5000);
            FetchDone();
        }

        public void FetchStart()
        {
            if (this.InvokeRequired)
            {
                this.Invoke(new Action(delegate()
                {
                    aTimer.Enabled = false;
                    picSlide.SizeMode = PictureBoxSizeMode.CenterImage;
                    picSlide.Image = SchedulerWithTimer.Properties.Resources.loading;
                }));
            }
            else
            {
                aTimer.Enabled = false;
                picSlide.SizeMode = PictureBoxSizeMode.CenterImage;
                picSlide.Image = SchedulerWithTimer.Properties.Resources.loading;
            }

        }

        public void FetchDone()
        {
            if (this.InvokeRequired)
            {
                this.Invoke(new Action(delegate()
                {
                    aTimer.Enabled = true;
                }));
            }
            else
                aTimer.Enabled = true;
        }


        private void OnTimedEvent(object source, ElapsedEventArgs e)
        {
            if (this.InvokeRequired)
            {
                this.Invoke(new Action(delegate()
                {
                    loadNextImage();
                }));
            }
            else
                loadNextImage();
        }

        private void loadNextImage()
        {
            if (this.InvokeRequired)
            {
                this.Invoke(new Action(delegate()
                {
                    if (imageNumber == SlideCount)
                    {
                        imageNumber = 1;
                    }
                    picSlide.SizeMode = PictureBoxSizeMode.Zoom;
                    picSlide.ImageLocation = string.Format(_SlideSource + @"\Slide{0}.jpg", imageNumber);

                }));
            }
            else
            {
                if (imageNumber == SlideCount)
                {
                    imageNumber = 1;
                }
                picSlide.SizeMode = PictureBoxSizeMode.Zoom;
                picSlide.ImageLocation = string.Format(_SlideSource + @"\Slide{0}.jpg", imageNumber);
                imageNumber++;
            }
        }
    }
}

may be some day my post may help other. thanks all :)
